lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -34,6 +37,70 @@ namespace circt {
 /// Calyx component.
 using FuncMapping = DenseMap<FuncOp, calyx::ComponentOp>;
 
+using Scheduleable = std::variant<calyx::GroupOp>;
+
+//===----------------------------------------------------------------------===//
+// Utility functions
+//===----------------------------------------------------------------------===//
+
+/// Tries to match a constant value defined by op. If the match was
+/// successful, returns true and binds the constant to 'value'.
+static bool matchConstantOp(Operation *op, APInt &value) {
+  return mlir::detail::constant_int_op_binder(&value).match(op);
+}
+
+/// Creates a DictionaryAttr containing a unit attribute 'name'. Used for
+/// defining mandatory port attributes for calyx::ComponentOp's.
+static DictionaryAttr getMandatoryPortAttr(MLIRContext *ctx, StringRef name) {
+  return DictionaryAttr::get(
+      ctx, {NamedAttribute(Identifier::get(name, ctx), UnitAttr::get(ctx))});
+}
+
+/// Adds the mandatory Calyx component I/O ports (->[clk, reset, go], [done]->)
+/// to ports
+static void addMandatoryComponentPorts(PatternRewriter &rewriter,
+                                       SmallVector<calyx::PortInfo> &ports) {
+  MLIRContext *ctx = rewriter.getContext();
+  ports.push_back({.name = rewriter.getStringAttr("clk"),
+                   .type = rewriter.getI1Type(),
+                   .direction = calyx::Direction::Input,
+                   .attributes = getMandatoryPortAttr(ctx, "clk")});
+  ports.push_back({.name = rewriter.getStringAttr("reset"),
+                   .type = rewriter.getI1Type(),
+                   .direction = calyx::Direction::Input,
+                   .attributes = getMandatoryPortAttr(ctx, "reset")});
+  ports.push_back({.name = rewriter.getStringAttr("go"),
+                   .type = rewriter.getI1Type(),
+                   .direction = calyx::Direction::Input,
+                   .attributes = getMandatoryPortAttr(ctx, "go")});
+  ports.push_back({.name = rewriter.getStringAttr("done"),
+                   .type = rewriter.getI1Type(),
+                   .direction = calyx::Direction::Output,
+                   .attributes = getMandatoryPortAttr(ctx, "done")});
+}
+
+/// Creates a new group within compOp.
+template <typename TGroup, typename TRet = TGroup>
+static TRet createGroup(PatternRewriter &rewriter, calyx::ComponentOp compOp,
+                        Location loc, Twine uniqueName) {
+
+  IRRewriter::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPointToEnd(compOp.getWiresOp().getBody());
+  auto groupOp = rewriter.create<TGroup>(loc, uniqueName.str());
+  rewriter.createBlock(&groupOp.getBodyRegion());
+  return groupOp;
+}
+
+/// Get the index'th output of compOp, which is associated with
+/// funcOp.
+
+static Value getComponentOutput(mlir::FuncOp funcOp, calyx::ComponentOp compOp,
+                                unsigned index) {
+  size_t resIdx = funcOp.getNumArguments() + 3 /*go, reset, clk*/ + index;
+  assert(compOp.getNumArguments() > resIdx);
```

Comment:
```suggestion
  assert(compOp.getNumArguments() > resIdx && "Exceeded number of arguments in the Component");
```

Also maybe we should document that this assumes the invariant that both the ComponentOp and FuncOp hope the same port ordering.